trl/trainer/grpo_trainer.py

```diff
@@ -1195,14 +1203,73 @@ def _calculate_rewards(self, inputs, prompts, completions, completion_ids_list):
+            else:
+                # Separate async and sync custom functions
+                if asyncio.iscoroutinefunction(reward_func):
+                    async_funcs_info.append((i, reward_func, reward_func_name))
                 else:
-                    output_reward_func = reward_func(
-                        prompts=prompts, completions=completions, completion_ids=completion_ids_list, **reward_kwargs
+                    sync_funcs_info.append((i, reward_func, reward_func_name))
+
+        # Execute non-async custom functions in parallel using ProcessPoolExecutor
+        if sync_funcs_info:
+
+            def _invoke_sync_reward(index, func, func_name):
+                with profiling_context(self, func_name):
+                    output = func(
+                        prompts=prompts,
+                        completions=completions,
+                        completion_ids=completion_ids_list,
+                        **reward_kwargs,
                     )
-                    # Convert None values to NaN
-                    output_reward_func = [reward if reward is not None else torch.nan for reward in output_reward_func]
+                    output = [r if r is not None else torch.nan for r in output]
+                    return index, output
+
+            with ProcessPoolExecutor(mp_context=multiprocessing.get_context("spawn")) as executor:
+                sync_futures = [
+                    executor.submit(_invoke_sync_reward, i, func, func_name)
+                    for (i, func, func_name) in sync_funcs_info
+                ]
+
+                for future in as_completed(sync_futures):
+                    idx, output_reward_func = future.result()
+                    rewards_per_func[:, idx] = torch.tensor(output_reward_func, dtype=torch.float32, device=device)
+
+        # Execute async custom functions in parallel using asyncio.gather
+        if async_funcs_info:
+
+            async def _invoke_async_reward(index, func, func_name):
+                with profiling_context(self, func_name):
+                    output = await func(
+                        prompts=prompts,
+                        completions=completions,
+                        completion_ids=completion_ids_list,
+                        **reward_kwargs,
+                    )
+                    output = [r if r is not None else torch.nan for r in output]
+                    return index, output
+
+            async def _run_async_funcs():
+                coros = [_invoke_async_reward(i, func, func_name) for (i, func, func_name) in async_funcs_info]
+                return await asyncio.gather(*coros)
+
+            # Detect if a loop is already running (e.g., pytest-asyncio). If so create a new loop to avoid
+            # creating coroutines that never get awaited due to fallback.
+            try:
+                running_loop = asyncio.get_running_loop()
+            except RuntimeError:
+                running_loop = None
```

Comment:
This may be a silly question, but I'm not very familiar with asyncio. How can we have a loop already running?